src/common/ZulipBanner.js

```diff
@@ -0,0 +1,91 @@
+/* @flow strict-local */
+
+import React from 'react';
+import type { Node } from 'react';
+import { View } from 'react-native';
+import { SafeAreaView } from 'react-native-safe-area-context';
+
+import { createStyleSheet, HALF_COLOR } from '../styles';
+import type { LocalizableText } from '../types';
+import Label from './Label';
+import ZulipTextButton from './ZulipTextButton';
+
+// `textRow` and `buttonsRow` are named for the more common case where
+// there's not enough room for the text and the button(s) to share a single
+// row.
+//
+// But we do support having the text and button(s) share a row if there's
+// room, i.e., if the entities' combined widths are less than or equal to
+// one row width. The spec gives an example of this in an illustration where
+// the text just takes one line and there's only one action button.
+//
+// TODO(?): The vertical centering logic for when the text and buttons share
+//   a row isn't pixel-perfect; it depends on explicit padding/margin values
+//   instead of declaring "center" somewhere. This is an intentional
+//   compromise to reduce complexity while still supporting the two-row
+//   layout. If we find an elegant solution, we should use it.
+const styles = createStyleSheet({
+  wrapper: {
+    flexDirection: 'row',
+    flexWrap: 'wrap',
+    backgroundColor: HALF_COLOR,
+    paddingLeft: 16,
+    paddingRight: 8,
+    paddingBottom: 8,
+    paddingTop: 10,
+  },
+  textRow: {
+    flexGrow: 1,
+    flexDirection: 'row',
+    marginBottom: 12,
+  },
+  text: {
+    marginTop: 6,
+    lineHeight: 20,
+  },
+  buttonsRow: {
+    flexGrow: 1,
+    flexDirection: 'row',
+    justifyContent: 'flex-end',
+  },
+});
+
+type Button = $ReadOnly<{|
+  id: string,
+  label: LocalizableText,
+  onPress: () => void,
+|}>;
+
+type Props = $ReadOnly<{|
+  visible: boolean,
+  text: LocalizableText,
+  buttons: $ReadOnlyArray<Button>,
+|}>;
+
+/**
+ * A banner that follows Material Design specifications as much as possible.
+ *
+ * See https://material.io/components/banners.
+ */
+// Please consult the Material Design doc before making layout changes, and
+// try to make them in a direction that brings us closer to the guidelines.
+export default function ZulipBanner(props: Props): Node {
+  const { visible, text, buttons } = props;
+
+  if (!visible) {
+    return null;
+  }
+
+  return (
+    <SafeAreaView mode="padding" edges={['right', 'left']} style={styles.wrapper}>
+      <View style={styles.textRow}>
+        <Label style={styles.text} text={text} />
+      </View>
+      <View style={styles.buttonsRow}>
+        {buttons.map(({ id, label, onPress }) => (
+          <ZulipTextButton key={id} label={label} onPress={onPress} />
```

Comment:
Is `key` needed here?

I think it shouldn't be -- the caller isn't likely to have a dynamic list of buttons, but even if they do, I don't think `ZulipTextButton` has any state that would make it matter to preserve the identity of which button is which across re-renders.

Reply:
Without it, I see the expected warning message:

![image](https://user-images.githubusercontent.com/22248748/133676685-634ef62f-ecbe-4ca3-84ef-4f41bd75bcab.png)

That warning does go away if I use the list index for the `key` though. Should I do that?

Reply:
Hmm, I see. Looks like what's happening is that the old code with two separate `<ZulipTextButton … />` expressions in the source resulted in two separate arguments to `React.createElement`, while this version with `{buttons.map(…)}` results in passing the list as a single argument to `React.createElement`. Here's [a demo](https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABACwKYBt1wBQEpEDeAUIogE6pQhlIA8AJjAG4B8tAhogB4C8BAjAF9EAejadeBAEzCxtEY1YBuIoKJFQkWAhQYsUvIRKIICAM5RE6RD0QBtfgBpEUgLorSFKjUQNmLY1ICdAA6AFt2AAdsbC58HhZEbEDSXwk-LlkA1NTcXDVSeUUWFTUgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=react&prettier=false&targets=&version=7.15.6&externalPlugins=&assumptions=%7B%7D) (starting from the handy link at the top of ["JSX In Depth"](https://reactjs.org/docs/jsx-in-depth.html).)

And then the warning is presumably driven by seeing that list of elements, and pretty reasonably figuring that it's dynamic and should use `key`.

In that case, let's go ahead and use keys like you have them. I wouldn't feel good about using the list index for the key -- that's such a well-known bad pattern that using it here to shut up a warning would look like hubris inviting a bug :wink: